src/coins.cpp

```diff
@@ -96,19 +93,14 @@ void AddCoins(CCoinsViewCache& cache, const CTransaction &tx, int nHeight, bool
 }
 
 bool CCoinsViewCache::SpendCoin(const COutPoint &outpoint, Coin* moveout) {
-    CCoinsMap::iterator it = FetchCoin(outpoint);
-    if (it == cacheCoins.end()) return false;
-    cachedCoinsUsage -= it->second.coin.DynamicMemoryUsage();
+    Modifier modifier(*this, outpoint);
+    Coin& coin = modifier.Modify();
+    bool already_spent = coin.IsSpent();
     if (moveout) {
-        *moveout = std::move(it->second.coin);
-    }
-    if (it->second.flags & CCoinsCacheEntry::FRESH) {
-        cacheCoins.erase(it);
-    } else {
-        it->second.flags |= CCoinsCacheEntry::DIRTY;
-        it->second.coin.Clear();
+        *moveout = std::move(coin);
     }
-    return true;
+    coin.Clear();
+    return !already_spent;
```

Comment:
I think this change (after #10537) would have prevented the inflation bug caused by #9049 + #10195, discussed in
the [CVE-2018-17144](https://bitcoincore.org/en/2018/09/20/notice/) full disclosure.